cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,293 @@
+ */
+void FlowLayout::setGeometry(const QRect &rect)
+{
+    QLayout::setGeometry(rect);
+    int left, top, right, bottom;
+    getContentsMargins(&left, &top, &right, &bottom);
+    QRect adjustedRect = rect.adjusted(+left, +top, -right, -bottom);
+    int availableWidth = qMax(adjustedRect.width(), getParentScrollAreaWidth());
+    int totalHeight = layoutRows(adjustedRect.x(), adjustedRect.y(), availableWidth);
+
+    QWidget *parentWidgetPtr = parentWidget();
+    if (parentWidgetPtr) {
+        parentWidgetPtr->setMinimumSize(availableWidth, totalHeight);
+    }
+}
+
+/**
+ * @brief Arranges items in rows based on the available width.
+ * @param originX The starting x-coordinate for the row layout.
+ * @param originY The starting y-coordinate for the row layout.
+ * @param availableWidth The available width to lay out items.
+ * @return The y-coordinate of the final row's end position.
+ */
+int FlowLayout::layoutRows(int originX, int originY, int availableWidth)
+{
+    QVector<QLayoutItem *> rowItems;
+    int currentXPosition = originX;
+    int currentYPosition = originY;
+
+    int rowWidth = 0;
+    int rowHeight = 0;
+
+    for (QLayoutItem *item : items) {
+        QSize itemSize = item->sizeHint();
+        int itemWidth = itemSize.width() + horizontalSpacing();
+
+        if (currentXPosition + itemWidth > availableWidth) {
+            layoutRow(rowItems, originX, currentYPosition);
+            rowItems.clear();
+            currentXPosition = originX;
+            currentYPosition += rowHeight + verticalSpacing();
+            rowWidth = 0;
+            rowHeight = 0;
+        }
+        rowItems.append(item);
+        rowWidth += itemSize.width() + horizontalSpacing();
+        rowHeight = qMax(rowHeight, itemSize.height());
+        currentXPosition += itemSize.width() + horizontalSpacing();
+    }
+
+    layoutRow(rowItems, originX, currentYPosition);
+
+    return currentYPosition + rowHeight;
+}
+
+/**
+ * @brief Helper function for arranging a single row of items within specified bounds.
+ * @param rowItems Items to be arranged in the row.
+ * @param x The x-coordinate for starting the row.
+ * @param y The y-coordinate for starting the row.
+ */
+void FlowLayout::layoutRow(const QVector<QLayoutItem *> &rowItems, int x, int y)
+{
+    for (QLayoutItem *item : rowItems) {
+        QSize itemMaxSize = item->widget()->maximumSize();
+        int itemWidth = qMin(item->sizeHint().width(), itemMaxSize.width());
+        int itemHeight = qMin(item->sizeHint().height(), itemMaxSize.height());
+        item->setGeometry(QRect(QPoint(x, y), QSize(itemWidth, itemHeight)));
+        x += itemWidth + horizontalSpacing();
+    }
+}
+
+/**
+ * @brief Returns the preferred size for this layout.
+ * @return The maximum of all item size hints as a QSize.
+ */
+QSize FlowLayout::sizeHint() const
+{
+    QSize size;
+    for (QLayoutItem *item : items) {
+        size = size.expandedTo(item->sizeHint());
```

Comment:
nullptr check (I know i've commented this a few times, might want to do that on the insert instead?